### PR TITLE
Add free time tag

### PR DESCRIPTION
**Pull request by donwong2308**
Added Free Time Tags

Format

- Prefix: ft/
- Mon-Sun:HHmm-HHmm (24hr format)
- Example: `add n/pallon p/91111111 e/a@a.com r/00-00-00 t/pallon b/11/11/1111 ft/Mon:1300-1400`
- Example: `edit [INDEX] ft/Mon:1300-1400 ft/Wed:1300-1400`

Notes:
- Supports multiple free time tags
- Screenshot:
image

### CODE REVIEW

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 1]
```diff
@@ -17,17 +22,11 @@ public class SampleDataUtil {
     public static Person[] getSamplePersons() {
         return new Person[] {
             new Person(new Name("Alex Yeoh"), new Phone("87438807"), new Email("alexyeoh@example.com"),
-                new RoomNumber("21-06-40"), new Telegram("alexYeoh"), new Birthday("03/02/2000")),
```
Comment (nyankoclaws) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:20]: Just confirming we removing most sample persons?

File: src/main/java/seedu/address/model/tag/FreeTimeTag.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Free Time Tag in Dormie.
+ */
+public class FreeTimeTag extends Tag {
+
+    public static final String MESSAGE_CONSTRAINTS = "Free Time Tag should be Mon-Sun:HHmm-HHmm (24hr format)";
+    public static final String VALIDATION_REGEX =
+            "^(Mon|Tue|Wed|Thu|Fri|Sat|Sun):(0[0-9]|1[0-9]|2[0-3])([0-5][0-9])-(0[0-9]|1[0-9]|2[0-3])([0-5][0-9])$";
```
Comment (larainezo) [inline src/main/java/seedu/address/model/tag/FreeTimeTag.java:12]: This is a good format to save as. Perhaps if we have extra time in future, we can try to further separate the day and time! :)

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -121,4 +128,31 @@ public static Email parseEmail(String email) throws ParseException {
...
+        if (!FreeTimeTag.isValidTagName(trimmedTag)) {
+            throw new ParseException(FreeTimeTag.MESSAGE_CONSTRAINTS);
+        }
+        return new FreeTimeTag(trimmedTag);
+    }
+
+    /**
+     * Parses {@code Collection<String> tags} into a {@code Set<Tag>}.
+     */
+    public static Set<FreeTimeTag> parseFreeTimeTags(Collection<String> tags) throws ParseException {
```
Comment (larainezo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:150]: Will need to change from HashSet to HashMap

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -121,4 +128,31 @@ public static Email parseEmail(String email) throws ParseException {
...
+        requireNonNull(tags);
+        final Set<FreeTimeTag> tagSet = new HashSet<>();
+        for (String tagName : tags) {
+            tagSet.add(parseFreeTimeTag(tagName));
```
Comment (larainezo) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:154]: Can we show the multiple tags for one day? Since a single day may have multiple free time periods.

File: src/main/java/seedu/address/logic/commands/EditCommand.java
Comment (larainezo) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:1]: We may need to create an additional command to add/remove free time by a specific day, instead of having to retype for all days when only trying to edit for one day.

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 5]
```diff
@@ -29,13 +30,15 @@ public class AddCommand extends Command {
...
+            + "[" + PREFIX_FREETIMETAG + "TAG]...\n"
             + "Example: " + COMMAND_WORD + " "
             + PREFIX_NAME + "John Doe "
             + PREFIX_PHONE + "98765432 "
             + PREFIX_EMAIL + "johnd@example.com "
             + PREFIX_ROOMNUMBER + "sw-01-01 "
             + PREFIX_TELEGRAM + "johnDoe "
-            + PREFIX_BIRTHDAY + "12/12/2000 ";
+            + PREFIX_BIRTHDAY + "12/12/2000 "
+            + PREFIX_FREETIMETAG + "Mon:1pm-2pm";
```
Comment (larainezo) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:41]: Maybe we can show for multiple days? To show the user that they can set timings for multiple days upon creation.

**Review by larainezo — APPROVED** [review]
Looks good! Great work!

File: src/main/java/seedu/address/model/person/Person.java [hunk 6]
```diff
@@ -117,20 +130,23 @@ public boolean equals(Object other) {
...
-        return Objects.hash(name, phone, email, roomNumber, telegram, birthday);
+        return Objects.hash(name, phone, email, roomNumber, telegram, birthday, tags);
     }
 
     @Override
     public String toString() {
         ToStringBuilder sb = new ToStringBuilder(this);
         sb.add("name", name);
         sb.add("phone", phone);
+        sb.add("tags", tags);
```
Comment (PallonCX) [inline src/main/java/seedu/address/model/person/Person.java:149]: I'm quite curious that when is this function used, I didn't manage to find any in our current codebase (except its corresponding test function).

My main concern is the formatting issue as tags is a Set instead of a string, and it will be good if there is some visible usage to verify whether it's behaviour is as intended.

File: src/main/java/seedu/address/storage/JsonAdaptedFreeTimeTag.java
Comment (PallonCX) [inline src/main/java/seedu/address/storage/JsonAdaptedFreeTimeTag.java:1]: May I know when will this class be used?

File: src/test/java/seedu/address/logic/LogicManagerTest.java [hunk 7]
```diff
@@ -169,7 +169,7 @@ public void saveAddressBook(ReadOnlyAddressBook addressBook, Path filePath)
         // Triggers the saveAddressBook method by executing an add command
         String addCommand = AddCommand.COMMAND_WORD + NAME_DESC_AMY + PHONE_DESC_AMY
                 + EMAIL_DESC_AMY + ROOMNUMBER_DESC_AMY + TELEGRAM_DESC_AMY + BIRTHDAY_DESC_AMY;
-        Person expectedPerson = new PersonBuilder(AMY).build();
+        Person expectedPerson = new PersonBuilder(AMY).withFreeTimeTags().build();
```
Comment (PallonCX) [inline src/test/java/seedu/address/logic/LogicManagerTest.java:172]: May I know why is this change needed? Looks like the constructor of PersonBuilder with the parameter AMY will already initialize the freeTimeTags?

**Review by PallonCX — COMMENTED** [review]
Thanks for your effort! Just wondering whether it is possible to make the increment smaller (i.e. less changes in a single pull request) and separate the changes to multiple pull requests? By doing so, I believe it will be easier to review as the dependency between files are decreased.

File: src/main/java/seedu/address/logic/Messages.java [hunk 8]
```diff
@@ -56,6 +56,11 @@ public static String format(Person person) {
             builder.append("; Birthday: ").append(person.getBirthday());
         }
 
+        if (!person.getTags().isEmpty()) {
+            builder.append("; Free Time Tags: ");
+            person.getTags().forEach(builder::append);
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/main/java/seedu/address/logic/Messages.java:61]: Yeap, added to the comments in the PR

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 9]
```diff
@@ -29,13 +30,15 @@ public class AddCommand extends Command {
             + PREFIX_ROOMNUMBER + "ROOM NUMBER "
             + PREFIX_TELEGRAM + "TELEGRAM "
             + PREFIX_BIRTHDAY + "BIRTHDAY "
+            + "[" + PREFIX_FREETIMETAG + "TAG]...\n"
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/main/java/seedu/address/logic/commands/AddCommand.java:33]: We can add that in the UG or the help command perhaps

File: src/main/java/seedu/address/logic/commands/AddCommand.java (hunk 5 above)
Comment (donwong2308) [reply src/main/java/seedu/address/logic/commands/AddCommand.java:41]: Yeap sounds good!

File: src/main/java/seedu/address/logic/commands/EditCommand.java
Comment (donwong2308) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:1]: That is a good point, we could create an issue for that

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 10]
```diff
@@ -44,6 +49,7 @@ public class EditCommand extends Command {
             + "[" + PREFIX_ROOMNUMBER + "ROOM NUMBER] "
             + "[" + PREFIX_TELEGRAM + "TELEGRAM] "
             + "[" + PREFIX_BIRTHDAY + "BIRTHDAY] "
+            + "[" + PREFIX_FREETIMETAG + "FREE TIME]...\n"
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:52]: We can add that in the UG or the help command perhaps

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 11]
```diff
@@ -101,9 +107,10 @@ private static Person createEditedPerson(Person personToEdit, EditPersonDescript
         RoomNumber updatedRoomNumber = editPersonDescriptor.getRoomNumber().orElse(personToEdit.getRoomNumber());
         Telegram updatedTelegram = editPersonDescriptor.getTelegram().orElse(personToEdit.getTelegram());
         Birthday updatedBirthday = editPersonDescriptor.getBirthday().orElse(personToEdit.getBirthday());
+        Set<FreeTimeTag> updatedTags = editPersonDescriptor.getTags().orElse(personToEdit.getTags());
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:110]: For now it is not possible, it is under the edit command

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 4 above)
Comment (donwong2308) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:154]: Yeap we can

File: src/main/java/seedu/address/model/person/Person.java (hunk 6 above)
Comment (donwong2308) [reply src/main/java/seedu/address/model/person/Person.java:149]: Agreed, perhaps we could look into this and see what are the effects of this

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 1 above)
Comment (donwong2308) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:20]: I was removing them for debugging purposes 😅
But as far as I see it the removal of it would not matter for now

File: src/main/java/seedu/address/storage/JsonAdaptedFreeTimeTag.java
Comment (donwong2308) [reply src/main/java/seedu/address/storage/JsonAdaptedFreeTimeTag.java:1]: If I'm not wrong its used in Storage to extract the stored tags by contact

File: src/test/java/seedu/address/logic/LogicManagerTest.java (hunk 7 above)
Comment (donwong2308) [reply src/test/java/seedu/address/logic/LogicManagerTest.java:172]: I added it while debugging as without it, it will cause the following test cases to fail as the expected/actual models do not match:
execute_storageThrowsIoException_throwsCommandException
execute_storageThrowsAdException_throwsCommandException

**Review by donwong2308 — COMMENTED** [review]
Address comments

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 1 above)
Comment (PallonCX) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:20]: Is it possible to add the test cases back? As I think it will always be better to have more test coverage, or will doing so cause any issue?

File: src/test/java/seedu/address/logic/LogicManagerTest.java (hunk 7 above)
Comment (PallonCX) [reply src/test/java/seedu/address/logic/LogicManagerTest.java:172]: Hmmm I thought `new PersonBuilder(AMY)` and `new PersonBuilder(AMY).withFreeTimeTags()` should behave the same? Or what are their difference?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 12]
```diff
@@ -121,4 +128,31 @@ public static Email parseEmail(String email) throws ParseException {
...
+        }
+        return tagSet;
+    }
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java:157]: I followed the AB3 structure for this, but I think the idea is that there will always exist a set of tags, it is just whether it will be empty or not

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 13]
```diff
@@ -5,6 +5,10 @@
 import static seedu.address.testutil.Assert.assertThrows;
 import static seedu.address.testutil.TypicalPersons.BENSON;
 
+//import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:8]: Thanks for pointing this out, will change it!

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 14]
```diff
@@ -14,6 +18,7 @@
 import seedu.address.model.person.Phone;
 import seedu.address.model.person.RoomNumber;
 import seedu.address.model.person.Telegram;
+//import seedu.address.model.tag.FreeTimeTag;
```
[the remark this answers is not in the document]
Comment (donwong2308) [reply src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:21]: Thanks for pointing this out, will change it!

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 1 above)
Comment (donwong2308) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:20]: Ok I'll add it back

**Review by PallonCX — APPROVED** [review]
LGTM

**Review by JustWeiHao — APPROVED** [review]
LGTM

**Comment by nyankoclaws** [comment]
LGTM

**Review by nyankoclaws — APPROVED** [review]
LGTM
